include/envoy/server/drain_manager.h

```diff
@@ -8,20 +8,17 @@ namespace Envoy {
 namespace Server {
 
 /**
- * Handles connection draining. An instance is generally shared across the entire server.
+ * Handles connection draining. This concept is used globally during hot restart / server draining
+ * as well as on individual listeners when they are being dynamically removed.
  */
 class DrainManager : public Network::DrainDecision {
 public:
   /**
-   * @return TRUE if the manager is currently draining connections.
+   * Invoked to begin the drain procedure. (Making drain close operations more likely).
+   * @param completion supplies the completion that will be called when the drain sequence is
+   *                   finished. The parameter is optional and can be an unassigned function.
    */
-  virtual bool draining() PURE;
-
-  /**
-   * Invoked in the secondary process to begin the drain procedure. (Making drain close operations
-   * more likely).
-   */
-  virtual void startDrainSequence() PURE;
+  virtual void startDrainSequence(std::function<void()> completion) PURE;
```

Comment:
Do we drain globally on each LDS update? This seems reasonable given the current semantics (each LDS refresh is an update of the complete set of listeners), but won't this also preclude the possibility of skipping the drain on listeners that aren't being modified (i.e. they have the same config hash)?